src/verify/mod.rs

```diff
@@ -0,0 +1,309 @@
+use std::collections::HashMap;
+
+use crate::ast;
+pub mod error;
+use error::Error;
+
+// Here, I'm using "item" to refer to things that exist in the language, for example variables,
+// function and soon types. Please change the name item to something better.
+//
+// During verification, every kind of item is represented by a struct containing the required
+// information.
+//
+// Then, a scope consists of a map from identifiers to items.
+// Currently there are 2 types of scope, `TopLevelScope` and `Scope`. These will soon just be one
+// however are separated for now to keep things simple as each kind of scope can only contain 1
+// kind of item (hence no need for an item enum yet).
+
+pub fn verify<'a>(items: &'a [ast::Item<'a>]) -> Vec<Error<'a>> {
+    let (top_level, mut errors) = TopLevelScope::build(items);
+    let more_errors = top_level.check_items(items);
+    errors.extend(more_errors);
+    errors
+}
+
+/// Represents a function. Currently, all we need for verification is the number of parameters.
+/// This will likely grow to include type and proof information for each parameter along with the
+/// return type.
+#[derive(Debug, Clone)]
+struct Func {
+    /// the number of parameters, when types are introduced this will likely be replaced with a
+    /// list of type information for each parameter.
+    n_params: usize,
+}
+
+/// Currently variables are all isize so no information is required!
```

Comment:
Minor correction: They're `usize`s. From [`ExprKind`](https://github.com/tc-lang/trenchcoat/blob/1423abd4912eb19622a57796564d14e5d94f399d/src/ast/mod.rs#L183):
```rust
pub enum ExprKind<'a> {
    // -- snip --

    /// A raw number, e.g. `420`
    Num(usize),

    // -- snip --
}
```

Reply:
These are variables though. Although constants are unsigned I think we decided that variables can have signed values.

Reply:
You're right - my bad!